src/ast/attachpoint_parser.cpp

```diff
@@ -0,0 +1,547 @@
+#include "attachpoint_parser.h"
+
+#include <algorithm>
+#include <exception>
+#include <functional>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "types.h"
+
+namespace {
+
+/*
+ * This function splits an attach point definition into arguments,
+ * where arguments are separated by `:`. The exception is `:`s inside
+ * of quoted strings, which we must treat as a literal.
+ *
+ * Note that this function assumes the raw string is generally well
+ * formed. More specifically, that there is no unescaped whitespace
+ * and no unmatched quotes.
+ */
+std::vector<std::string> split_attachpoint(const std::string &raw,
+                                           bool remove_empty = false)
+{
+  std::vector<std::string> ret;
+  bool in_quotes = false;
+  std::string argument;
+
+  for (char c : raw)
+  {
+    if (c == ':' && !in_quotes)
+    {
+      if (argument.empty() && remove_empty)
+        continue;
+
+      ret.emplace_back(std::move(argument));
+      // The standard says an std::string in moved-from state is in
+      // valid but unspecified state, so clear() to be safe
+      argument.clear();
+    }
+    else if (c == '"')
+      in_quotes = !in_quotes;
```

Comment:
I found the case an escaped quote is needed. e.g.,

```c
#include <stdio.h>
#include <sys/sdt.h>

int main(){
    int a = 1;
    DTRACE_PROBE1(test, "probe1", a);
    printf("%d\n", a);
    return 0;
}
```
then
```
% readelf -n a.out
[...]
Displaying notes found in: .note.stapsdt
  Owner                 Data size       Description
  stapsdt              0x00000032       NT_STAPSDT (SystemTap probe descriptors)
    Provider: test
    Name: "probe1"
    Location: 0x0000000000000659, Base: 0x0000000000000708, Semaphore: 0x0000000000000000
    Arguments: -4@-4(%rbp)
```
```sh
% sudo ./src/bpftrace -e 'usdt:./a.out:test:"\"probe1\"" { printf("%d\n", arg0); }'
couldn't get argument 0 for ./a.out:test:probe1
```
This works on master. I'm not sure but how about split attachpoint in parser?

Reply:
In this case, the string to parse is `usdt:./a.out:test:""probe1""`, and the follwoing is `usdt:./a.out:test:""pro:be1""`.
 
```sh
% sudo ./src/bpftrace -e 'usdt:./a.out:test:"\"pro:be1\"" { printf("%d\n", arg0); }'
stdin:1:1-32: ERROR: usdt probe type requires 2 or 3 arguments
usdt:./a.out:test:"\"pro:be1\"" { printf("%d\n", arg0); }
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Maybe `split_attachpoint` should only remove the first and the last quotes.

Reply:
Ah nice find.

> Maybe split_attachpoint should only remove the first and the last quotes.

Hm, that feels a bit brittle. I think lexers typically walk one character at a time and can peek/consume forward as necessary. Let me see if that will work here.